src/Data/List.purs

```diff
@@ -632,18 +637,65 @@ tails list@(Cons _ tl)= list : tails tl
 --------------------------------------------------------------------------------
 
 -- | Remove duplicate elements from a list.
+-- | Keeps the first duplicates found and preserves ordering otherwise.
+-- |
+-- | ```purescript
+-- | nub 1:2:1:3:3:Nil = 1:2:3:Nil
+-- | ```
+-- |
+-- | Running time: `O(n log n)`
+nub :: forall a. Ord a => List a -> List a
+nub = nubBy compare
+
+-- | Remove duplicate elements from a list based on the provided ordering function.
+-- | Keeps the first duplicates found and preserves ordering otherwise.
+-- |
+-- | ```purescript
+-- | nubBy (compare `on` Array.length) ([1]:[2]:[3,4]:Nil) == [1]:[3,4]:Nil
+-- | ```
+-- |
+-- | Running time: `O(n log n)`
+nubBy :: forall a. (a -> a -> Ordering) -> List a -> List a
+nubBy p =
+  -- Add indices so we can recover original order after deduplicating.
+  addIndexReverse
+  -- Sort by original values to cluster duplicates.
+  >>> sortBy (p `on` snd)
+  -- Removing neighboring duplicates.
+  >>> nubByAdjacentReverse (\a b -> (p `on` snd) a b == EQ)
```

Comment:
If we add an `Eq` constraint, this could be simplified to:
```purs
nubByAdjacentReverse (eq `on` snd)
```
That would probably also be faster, but seems to be against the spirit of allowing a custom ordering predicate.